provider/vsphere/internal/vsphereclient/client.go

```diff
@@ -0,0 +1,327 @@
+				return finder.Folder(ctx, parent.InventoryPath+"/"+name)
+			}
+		}
+		return folder, err
+	}
+
+	parentFolder := folders.VmFolder
+	for _, name := range strings.Split(folderPath, "/") {
+		folder, err := createFolder(parentFolder, name)
+		if err != nil {
+			return errors.Annotatef(
+				err, "creating folder %q in %q",
+				name, parentFolder.InventoryPath,
+			)
+		}
+		parentFolder = folder
+	}
+	return nil
+}
+
+// DestroyVMFolder destroys a folder rooted at the datacenter's base VM folder.
+func (c *Client) DestroyVMFolder(ctx context.Context, folderPath string) error {
+	finder, datacenter, err := c.finder(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	folders, err := datacenter.Folders(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	folderPath = path.Join(folders.VmFolder.InventoryPath, folderPath)
+	folder, err := finder.Folder(ctx, folderPath)
+	if err != nil {
+		if _, ok := err.(*find.NotFoundError); ok {
+			return nil
+		}
+		return errors.Trace(err)
+	}
+
+	task, err := folder.Destroy(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if _, err := task.WaitForResult(ctx, nil); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// MoveVMFolderInto moves one VM folder into another.
+func (c *Client) MoveVMFolderInto(ctx context.Context, parentPath, childPath string) error {
+	finder, datacenter, err := c.finder(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	folders, err := datacenter.Folders(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	parentPath = path.Join(folders.VmFolder.InventoryPath, parentPath)
+	childPath = path.Join(folders.VmFolder.InventoryPath, childPath)
+	parent, err := finder.Folder(ctx, parentPath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	child, err := finder.Folder(ctx, childPath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	task, err := parent.MoveInto(ctx, []types.ManagedObjectReference{child.Reference()})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if _, err := task.WaitForResult(ctx, nil); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// MoveVMsInto moves a set of VMs into a folder.
+func (c *Client) MoveVMsInto(
+	ctx context.Context,
+	folderPath string,
+	vms ...types.ManagedObjectReference,
+) error {
+	finder, datacenter, err := c.finder(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	folders, err := datacenter.Folders(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	folderPath = path.Join(folders.VmFolder.InventoryPath, folderPath)
+	folder, err := finder.Folder(ctx, folderPath)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	task, err := folder.MoveInto(ctx, vms)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if _, err := task.WaitForResult(ctx, nil); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// UpdateVirtualMachineExtraConfig updates the "ExtraConfig" attributes
+// of the specified virtual machine. Keys with empty values will be
```

Comment:
Not at the moment, and probably never. If we find we need to, we can change this internal API.